### PR TITLE
DCS-335 Add rest endpoints for adding/updating/deleting functional mailboxes

### PR DESCRIPTION
**Pull request by bell-pa**
* Add PUT and DELETE for functional mailbox at LDU and Team level
* Remove end-points identified by LDU only.  An LDU is now identified by (Probation Area Code, Local Delivery Unit Code) -> drop code + table
* Convert Java code to Kotlin
* Gradle build file is now in Kotlin
* Uses JUnit 5 + Mockk

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    kotlin("jvm") version "1.3.61"
+
+    id("io.spring.dependency-management") version "1.0.9.RELEASE"
+    id("org.springframework.boot") version "2.2.4.RELEASE"
+
+    // Makes classes annotated with @Component, @Async, @Transactional, @Cacheable and @SpringBootTest open
+    kotlin("plugin.spring") version "1.3.61"
+
+    // Adds a no-arg (Java) constructor to classes annotated with @Entity, @Embeddable or @MappedSuperclass
+    kotlin("plugin.jpa") version "1.3.61"
```
Comment (andrewrlee) [inline build.gradle.kts]: fancy!

File: src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+@Api(tags = ["probation-areas"])
+@RestController
+@RequestMapping(
+        value = ["probation-areas"],
+        produces = [APPLICATION_JSON_VALUE])
+@Slf4j
+class ProbationAreaController(val localDeliveryUnitService: LocalDeliveryUnit2Service) {
+
+    @GetMapping(
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt]: If we do a UI for this we might need to retrieve all LDUs for probation areas

File: src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
+import io.swagger.annotations.ApiModel
+import io.swagger.annotations.ApiModelProperty
+import javax.validation.constraints.Email
+import javax.validation.constraints.NotBlank
+
+@ApiModel(description = "Local Delivery Unit")
+data class LocalDeliveryUnit2Dto @JsonCreator constructor(
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt]: Not part of this PR, guess this could do with updating now we've got  rid of the original version

File: src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        @ApiModelProperty(required = true, value = "Probation Area code", position = 1, example = "NO2")
+        @JsonProperty("probationAreaCode")
+        val probationAreaCode: @NotBlank String,
+
+        @ApiModelProperty(required = true, value = "Local Delivery Unit code", position = 1, example = "NO2SUK")
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt]: could potentially add `springfox-bean-validators` plugin which pulls required-ness from @NotBlank annotations (and supports other validations) meaning you don't need to repeat in swagger docs 

Guess max string length and things like that could be enforced if appropriate too.

File: src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        assertThat(persistentLdu.id).isNotNull()
+
+        assertThat(persistentLdu.probationTeams).isEqualTo(lduWithProbationTeams(lduCode).probationTeams)
+
+        val id = persistentLdu.id
+        if (id != null) {
+            assertThat(probationTeamCount(id)).isEqualTo(2)
+        } else {
+            fail("No ID for persisted LDU")
+        }
```
Comment (andrewrlee) [inline src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt]: rather than having this style of logic in tests could use standard assertions: `assertThat(persistentLdu.id).isNotNull()`

File: src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            assertThat(probationTeamCount(lduId)).isEqualTo(2)
+            repository.deleteById(lduId)
+            TestTransaction.flagForCommit()
+            TestTransaction.end()
+
+            TestTransaction.start()
+
+            assertThat(lduCount(lduId)).isEqualTo(0)
+            assertThat(probationTeamCount(lduId)).isEqualTo(0)
+        } else {
```
Comment (andrewrlee) [inline src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt]: again I think all these could use `assertThat(blah).isNotNull()`

File: src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt (hunk 2 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt]: Wouldn't you ask Delius what the Probation Areas, LDUs were?  This service doesn't have that information.

File: src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt (hunk 3 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt]: Good point

File: src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt (hunk 4 above)
Comment (bell-pa) [reply src/main/java/uk/gov/justice/hmpps/probationteams/dto/LocalDeliveryUnit2Dto.kt]: I didn't know about that.  Separate PR? Tech debt?

File: src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt (hunk 5 above)
Comment (bell-pa) [reply src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt]: Yes

File: src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt (hunk 2 above)
Comment (andrewrlee) [reply src/main/java/uk/gov/justice/hmpps/probationteams/controllers/ProbationAreaController.kt]: Yes you would to get the relationship from delius. But you may want to decorate that info with info about mailboxes and to do that you'd want to bulk load of whatever infos available for a probation area might be good. 

(Still we'd only add it if we need it )
